File: Leetcode/Algorithm/cpp/00107-Binary Tree Level Order Traversal II.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        vector<vector<int> > ans;
        if (!root) {
            return ans;
        }
        
        queue<TreeNode*> q[2];
        int now = 0;
        
        q[now].push(root);
        while (!q[now].empty()) {
            int next = now ^ 1;
            q[next] = queue<TreeNode*>();
            ans.push_back(vector<int>());
            vector<int>& vec = *(--ans.end());
            while (!q[now].empty()) {
                TreeNode* p = q[now].front();
                q[now].pop();
                if (p -> left) {
                    q[next].push(p -> left);
                }
                if (p -> right) {
                    q[next].push(p -> right);
                }
                
                vec.push_back(p -> val);
            }
            now = next;
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

File: Leetcode/Algorithm/python/0000/00107-Binary Tree Level Order Traversal II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrderBottom(self, root):
        if not root:
            return []
        res = []
        now_level = [root]
        next_level = []
        while now_level:
            for node in now_level:
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
            res.append(map(lambda x: x.val, now_level))
            now_level = next_level
            next_level = []
        return res[::-1]
